### ISSUE TITLE
tfjs_binding.node not found, following instructions does nothing

### ISSUE THREAD

**Issue by CarlFink** [fangfufu/Linux-Fake-Background-Webcam#49]
I'm trying to set up on a Debian Buster system using Intel graphics. Here's what happens when I try to start the app:

carlf@debian-NUCi5:~/Linux-Fake-Background-Webcam/bodypix$ nodejs app.js 
No GPU found at /dev/nvidia0, using CPU
/home/carlf/Linux-Fake-Background-Webcam/bodypix/node_modules/@tensorflow/tfjs-node/dist/index.js:49
    throw new Error("The Node.js native addon module (tfjs_binding.node) can not "
    ^

Error: The Node.js native addon module (tfjs_binding.node) can not be found at path: /home/carlf/Linux-Fake-Background-Webcam/bodypix/node_modules/@tensorflow/tfjs-node/lib/napi-v5/tfjs_binding.node. 
Please run command 'npm rebuild @tensorflow/tfjs-node build-addon-from-source' to rebuild the native addon module. 
If you have problem with building the addon module, please check https://github.com/tensorflow/tfjs/blob/master/tfjs-node/WINDOWS_TROUBLESHOOTING.md or file an issue.
    at Object.<anonymous> (/home/carlf/Linux-Fake-Background-Webcam/bodypix/node_modules/@tensorflow/tfjs-node/dist/index.js:49:11)
    at Module._compile (internal/modules/cjs/loader.js:1236:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1257:10)
    at Module.load (internal/modules/cjs/loader.js:1085:32)
    at Function.Module._load (internal/modules/cjs/loader.js:950:14)
    at Module.require (internal/modules/cjs/loader.js:1125:19)
    at require (internal/modules/cjs/helpers.js:75:18)
    at tensorflow (/home/carlf/Linux-Fake-Background-Webcam/bodypix/app.js:45:16)
    at Object.<anonymous> (/home/carlf/Linux-Fake-Background-Webcam/bodypix/app.js:3:12)
    at Module._compile (internal/modules/cjs/loader.js:1236:30)

When I try to follow the instructions about compiling the missing module, that fails. I am not familiar at all with node.js and only very slightly with JavaScript, so I can't debug the resulting errors. The output is fairly long, so I captured it with script and I am uploading the typescript as an attachment to this report. The upshot is that compilation fails. 

I do notice the line:

In file included from ../binding/tfjs_backend.cc:18:
../binding/tfjs_backend.h:26:10: fatal error: tensorflow/c/eager/c_api.h: No such file or directory

However, `locate tensorflow | grep c_api` results in:
carlf@debian-NUCi5:~/Linux-Fake-Background-Webcam/bodypix$ locate tensorflow | grep c_api
/home/carlf/Linux-Fake-Background-Webcam/bodypix/node_modules/@tensorflow/tfjs-node/deps/include/tensorflow/c/c_api.h
/home/carlf/Linux-Fake-Background-Webcam/bodypix/node_modules/@tensorflow/tfjs-node/deps/include/tensorflow/c/c_api_experimental.h

Is the script looking in the wrong place? The file is on my system but not in the directory mentioned in the error message.

Thank you. I hope to eventually be able to try your seemingly-wonderful creation.
[typescript.txt](https://github.com/fangfufu/Linux-Fake-Background-Webcam/files/4975484/typescript.txt)

**Comment by fangfufu** [comment]
Please clone a fresh repository, run the following:

    cd bodypix
    npm install

Please post the output in a code block. It is hard to figure out what's going out during the installation process.

**Comment by CarlFink** [comment]
> Please clone a fresh repository, run the following:
> 
> ```
> cd bodypix
> npm install
> ```
> 
> Please post the output in a code block. It is hard to figure out what's going out during the installation process.

Thanks for getting back so fast (on a weekend!).

Again, I captured the result using script. Here you go:

`carlf@debian-NUCi5:~/Linux-Fake-Background-Webcam/bodypix$ npm install

> @tensorflow/tfjs-node@1.7.4 install /home/carlf/Linux-Fake-Background-Webcam/bodypix/node_modules/@tensorflow/tfjs-node
> node scripts/install.js

CPU-linux-1.7.4.tar.gz
* Downloading libtensorflow
[==============================] 7201561/bps 100% 0.0s
* Building TensorFlow Node.js bindings

> @tensorflow/tfjs-node-gpu@1.7.4 install /home/carlf/Linux-Fake-Background-Webcam/bodypix/node_modules/@tensorflow/tfjs-node-gpu
> node scripts/install.js gpu download

GPU-linux-1.7.4.tar.gz
* Downloading libtensorflow
[==============================] 5548712/bps 100% 0.0s
* Building TensorFlow Node.js bindings
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN bodypix@0.0.1 No description
npm WARN bodypix@0.0.1 No repository field.
npm WARN bodypix@0.0.1 No license field.

added 95 packages from 100 contributors and audited 95 packages in 40.876s
1 package is looking for funding
  run `npm fund` for details

found 0 vulnerabilities

carlf@debian-NUCi5:~/Linux-Fake-Background-Webcam/bodypix$`

**Comment by fangfufu** [comment]
Erm, did it work?

**Comment by CarlFink** [comment]
> Erm, did it work?

I was just coming back to say that yes, now I can start the app!

But starting fake.py fails, with this message:

```
carlf@debian-NUCi5:~/Linux-Fake-Background-Webcam/fakecam$ python3 fake.py 
Cannot set camera property 3 to 1280, used value: 0.0
Cannot set camera property 4 to 720, used value: 0.0
Cannot set camera property 5 to 30, used value: 0.0
Traceback (most recent call last):
  File "fake.py", line 379, in <module>
    main()
  File "fake.py", line 367, in main
    use_akvcam=args.akvcam)
  File "fake.py", line 114, in __init__
    self.fake_cam = pyfakewebcam.FakeWebcam(v4l2loopback_path, self.width, self.height)
  File "/home/carlf/.local/lib/python3.7/site-packages/pyfakewebcam/pyfakewebcam.py", line 54, in __init__
    fcntl.ioctl(self._video_device, _v4l2.VIDIOC_S_FMT, self._settings)
OSError: [Errno 22] Invalid argument
```

**Comment by fangfufu** [comment]
You have to set your webcams properly using:

    -w WEBCAM_PATH, --webcam-path WEBCAM_PATH
                        Set real webcam path
    -v V4L2LOOPBACK_PATH, --v4l2loopback-path V4L2LOOPBACK_PATH
                        V4l2loopback device path

**Comment by CarlFink** [comment]
Thanks for the tip. I know that the video stream is /dev/video1. I have no idea what the V4L2LOOPBACK_PATH should be set to. There is nothing under /dev/v4l except the two empty directories, by-id and by-path.

**Comment by fangfufu** [comment]
You probably haven't even installed v4l2loopback properly. You should have multiple devices in /dev/video*.

**Comment by CarlFink** [comment]
I have 3. The raw video is /dev/video0. I am using webcamoid to preprocess it, and that stream is /dev/video1. Ideally, I'd be able to then hook fake.py after webcamoid to do the background. 

If /dev/video1 is the V4L2LOOPBACK_PATH, then what would be the WEBCAM_PATH? Or is V4L2LOOPBACK_PATH the *output* or target device, not an input? I'm afraid the syntax is opaque to me. Is one of those a device that fake.py creates, as opposed to one it looks for?

**Comment by fangfufu** [comment]
V4L2LOOPBACK_PATH should be /dev/video2. I have no idea. The real webcam tends to tie up 2 devices.

**Comment by CarlFink** [comment]
OK, I tried a bunch of combinations. This loaded without error:

`python3 fake.py  -w /dev/video0 -v /dev/video0`

That works! The /dev/video0 stream, as read by VLC, has the White House image in the background. And a terrible, terrible flicker.

Weirdly, Webcamoid now sees video0 as just the White House press room scene. VLC sees that same device as me superimposed on the horribly flickering bckground. I would intuitively have expected all software to see video0 as the same stream, but I guess not.

Hmm ... dropping both the fps setting of fake.py and Webcamoid makes the flickering far worse. That's also unintuitive.

Looks as if my i5-based Intel system just doesn't have the power to do non-chromakey virtual background, maybe?

Thanks again for helping.

**Comment by fangfufu** [comment]
Your ``-w`` and ``-v`` should be different devices. You are clearly doing something wrong. 

Also, I have a quad-core i7, and I am only getting about 10fps.

**Comment by CarlFink** [comment]
It might be more generous to say that something is configured wrong, instead of, y'know, making an accusation. I followed the documentation as much as possible, but I've never used node.js or Python-controlled video. Debugging it myself would take more than the hour I've spent on this project.

Thank you for the info about CPU-bounding of the program. Looks like I'll just do without background removal for the time being.
